Introduce guard_or_true, guard_or_false

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #148430

some context in this document:
https://docs.google.com/document/d/18nJsj-F2C_QXO7ClwzPcAUENQ-B440B43W7DdDnlDt4/edit?tab=t.0#heading=h.pgebnyi7pocj

But TLDR;
`guard_or_true`, `guard_or_false` are better than `guard_size_oblivious` due to :
- Easier to reason about what assumptions we are making while reading the code.
- Avoid size_oblivious complexity that is not needed.
- Avoid unsoundness that could make `guard_size_oblivious(a==1)` be true when its not true for some vaue `a` during runtime.
- Less data dependent errors for some cases: ex, when doing `guard_size_oblivious(a==1)` and we know `a` is a tensor size, if it's traced with `a=u1-u2` `guard_size_oblivious(a==1)` will throw a data dependent error but `guard_else_false` will just return `False`.

### How is it different from statically_known_true??
**`if(cond)`:** (normal guarding) will try to evaluate statically and guard on the condition, willing to restrict input space to evaluate cond. if it fails to evaluate due to data dependent error will throw an exception (that could be converted to graph break in some situations). 

**`statically_known_true(cond)`:** would be used when you never want to add a guard (restrict your input space), but just want to do a best effort check to see if you can infer that something is true/false ONLY based on existing constraints.

**`guard_or_true(cond)`/`guard_or_false(cond)`:** Those would be used in situations you prefer to guard and know the result of the expression over not guarding, but in case you hit a data dependent error you are ok with just returning true or false.
Some reasons you might be ok with returning true/false instead could be:
1. It's an optimization I do not want to fail for not performing optimization.
2. I am willing to deviate from the normal semantics when I have unbacked for the benefit of not failing (See the doc above for more details).

**`definitely_true(cond)`**: same as `guard_or_false(cond)` except does not try to do static eval for unbacked (planning to deprecate it and replace uses with `guard_or_false` or make it alias to `guard_or_false`)
 


cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

laithsakka committed:
add gaurd else true, guard else false and gaurd_size_oblivious in decompositions.py

[ghstack-poisoned]

laithsakka committed:
Update on "Introduce guard_else_true, guard_else_false and avoid guard_size_oblivious in decompositions.py"


some context in this document:
https://docs.google.com/document/d/18nJsj-F2C_QXO7ClwzPcAUENQ-B440B43W7DdDnlDt4/edit?tab=t.0#heading=h.pgebnyi7pocj

But TLDR; 
guard_else_true, guard_else_false are better than guard_size_oblivious due to :
1) easier to reason about what assumptions we are making while reading the code. 
2) avoid guard_size_oblivious complexity that is not needed. 
3) avoid unsoundness that could make guard_size_oblivious(a==1) be True when its not True for some a during runtime. 
4) less data dependent errors for some cases: ex, when  guard_size_oblivious(a==1)  we know a is a tensor size, but its traced with a=u1-u2 for example 
guard_size_oblivious(a==1)  will throw data dependent error but guard_else_false will just return false. 

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Introduce guard_else_true, guard_else_false and avoid guard_size_oblivious in decompositions.py"


some context in this document:
https://docs.google.com/document/d/18nJsj-F2C_QXO7ClwzPcAUENQ-B440B43W7DdDnlDt4/edit?tab=t.0#heading=h.pgebnyi7pocj

But TLDR; 
guard_else_true, guard_else_false are better than guard_size_oblivious due to :
1) easier to reason about what assumptions we are making while reading the code. 
2) avoid guard_size_oblivious complexity that is not needed. 
3) avoid unsoundness that could make guard_size_oblivious(a==1) be True when its not True for some a during runtime. 
4) less data dependent errors for some cases: ex, when  guard_size_oblivious(a==1)  we know a is a tensor size, but its traced with a=u1-u2 for example 
guard_size_oblivious(a==1)  will throw data dependent error but guard_else_false will just return false. 

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

bobrenjc93 commented:
I instinctively don't like the named `guard_else_false` and `guard_else_true`. I have a much stronger preference for `guard_or_false` and `guard_or_true` since 1) there is already precedence for this type of wording in libraries like Rust's option type (https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or) and 2) it's less verbose 3) it reads more naturally to me. I think we've discussed this a few times, but since we are nearing landing the first PRs, I'd much rather we have the discussion explicitly than implicitly. Do you feel strongly about the `else` over `or`? If not can we please change to use the "road more traveled" naming for this type of API?

I do think we want to get buy in from our partners in dynamo/inductor before landing a change like this. As we discussed yesterday, there is nothing more permanent than a temporary commit and this is doubly true for public API changes. I'm adding @jansel, @eellison, @zou3519 and @anijain2305 as reviewers but feel free to add others that you think might have opinions.

eellison commented on torch/_inductor/decomposition.py:
```diff
@@ -315,21 +317,21 @@ def mm(
-            and guard_size_oblivious(self.size(0) > 0)
-            and guard_size_oblivious(input2.size(0) == 1)
+            guard_else_false(self.size(-1) == 1)
+            and guard_else_true(self.size(0) > 0)
+            and guard_else_false(input2.size(0) == 1)
```
Shouldn't we just be using statically_known_true ?

laithsakka replied:
statically_known_true will try to evaluate this statically, but it will fail in situations of unbacked for example
for the check guard_else_false(input2.size(0) == 1) when self.size(0) is u0 [0, infinity], in this case,
(1)  we can not tell statically if u0 ==1 or not. 
(2) we can not guard also on u0 being 1, because we do not have a hint. 

so what we do, we say we are willing to deviate here from the normal pytorch semantic for the runtime input of  input2.size(0)==1 and always take the false branch. 
We used to do that indirectly using guard_size_oblivious this is more explicit
see this document for more information about the intuition of guard_size_oblivious
https://docs.google.com/document/d/18nJsj-F2C_QXO7ClwzPcAUENQ-B440B43W7DdDnlDt4/edit?tab=t.0#heading=h.pgebnyi7pocj

eellison replied:
The decompositions here are not pytorch semantics. they are optimizations as decompositions.  

If `guard_else_false(input2.size(0) == 1) ` can be `statically_known_true(input2.size(0) == 1)`.

laithsakka replied:
i see do you want decompositions to fail if input2.size(0)  is unbacked?
if thats the case we can address that in different PR, since this PR does not change current semantics
but what you are proposing is to avoid (the avoidance of 0/1 specialization in this case)

eellison replied:
It's not very compelling to add two new apis which are only used in situations where we shouldn't be using them.

laithsakka replied:
I see so in those cases we actually want to fail with data dependent error if we hit those checks with unabcked? if thats the case i can use another example for this PR and migrate those to use normal conditions

eellison commented on torch/_inductor/decomposition.py:
```diff
@@ -261,13 +265,13 @@ def bmm(
     batch2: torch.Tensor,
 ) -> torch.Tensor:
     if config.coordinate_descent_tuning and self.device.type != "cpu":
-        if guard_size_oblivious(self.shape[1] == 1) or guard_size_oblivious(
+        if guard_else_false(self.shape[1] == 1) or guard_else_false(
             batch2.shape[2] == 1
         ):
             out = (self.unsqueeze(-1) * batch2.unsqueeze(1)).sum(dim=2)
             return out
     if self.device.type == "cpu":
-        if guard_size_oblivious(self.size(1) == 1) and guard_size_oblivious(
+        if guard_else_false(self.size(1) == 1) and guard_else_false(
             batch2.size(-1) == 1
         ):
```
Since these are all optimizations, we should just be doing statically_known_true. or potentially an api that is only willing to add guards on non-export use cases.

laithsakka replied:
mhmm this mostly keep the semantics of guard_size_oblivious, 
moving to statically_known_true will change the behaviour. 
i will start failing with unbacked.

this API does what you explained, just for unbacked vs backed instead of export vs non export.

bobrenjc93 replied:
@laithsakka I think what @eellison is saying is that in these cases we *should* change the behavior and move to `statically_known_true` even for the case of unbacked. Because these are optimizations, we don't want to guard, even if we can.

laithsakka replied:
i see make sense we can address that in different PR maybe? or I can put PR for that and rebase this on top of it .

eellison reviewed (commented):
not clear to me what the overlap is of this versus statically_known_true. even in this pr we have some wrong usages. i think we should be careful about adding more apis.

I also dont like the name guard_else_false and guard_else_true.

laithsakka commented:
i initially did not have strong opinion about the naming but as i implemented it i found guard_or_false, does not make sense,

first guard_or_false is x . logically
and if was meant (for either guard or false) I still fell that guard_else_false is better because it impose ordering
guard_otherwise_false works for me also.
the ordering is you try to guard first then if you fail you return false.

bobrenjc93 commented:
> first x_or_false is x . logically

Can you elaborate?

In python

>>> None or False
False

In javascript

>>> null || false
false

Also FWIW chatgpt also concurs:
<img width="1003" alt="Screenshot 2025-03-04 at 10 30 38 AM" src="https://github.com/user-attachments/assets/1ccb0a2e-d9ed-41b8-ac85-be453789f222" />

laithsakka commented:
> > > first x_or_false is x . logically
> > 
> > 
> > Can you elaborate?
> > In python
> > ```
> > >>> None or False
> > False
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In javascript
> > ```
> > >>> null || false
> > false
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Also FWIW chatgpt also concurs: <img alt="Screenshot 2025-03-04 at 10 30 38 AM" width="1003" src="https://private-user-images.githubusercontent.com/1324201/419145432-1ccb0a2e-d9ed-41b8-ac85-be453789f222.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDExMTM1NDgsIm5iZiI6MTc0MTExMzI0OCwicGF0aCI6Ii8xMzI0MjAxLzQxOTE0NTQzMi0xY2NiMGEyZS1kOWVkLTQxYjgtYWM4NS1iZTQ1Mzc4OWYyMjIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMDRUMTgzNDA4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWVkMjczMmNkOGY4NWZhOWEzYTNjYjFmN2M2NDdiZjAzZDBlMGFkYjMzNGQyMmY2NTE1Y2Y2OTdmOGIxODY3YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.qpaZgvyGVjM_qoN27n86NmEfmJ5WdKX0CPwKwJ5VF2I">
> 
> yeh but this is not null_or_false this guard_or_false? where guard has side effects.

if others agree i dont mind changing the name .

bobrenjc93 commented:
> where guard has side effects.

Guard only has side effects in the case where we guard and return True.

laithsakka committed:
Update on "Introduce guard_or_true, guard_or_false and avoid guard_size_oblivious in decompositions.py"


some context in this document:
https://docs.google.com/document/d/18nJsj-F2C_QXO7ClwzPcAUENQ-B440B43W7DdDnlDt4/edit?tab=t.0#heading=h.pgebnyi7pocj

But TLDR; 
guard_else_true, guard_else_false are better than guard_size_oblivious due to :
1) easier to reason about what assumptions we are making while reading the code. 
2) avoid size_oblivious complexity that is not needed. 
3) avoid unsoundness that could make guard_size_oblivious(a==1) be True when its not True for some a during runtime. 
4) less data dependent errors for some cases: ex, when  guard_size_oblivious(a==1)  we know a is a tensor size, but its traced with a=u1-u2 for example 
guard_size_oblivious(a==1)  will throw data dependent error but guard_else_false will just return false. 

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison commented:
I don't think we should have both true and false versions of this. We only have statically_known_true. Not also statically_known_false.

laithsakka committed:
Update on "Introduce guard_or_true, guard_or_false and avoid guard_size_oblivious in decompositions.py"


some context in this document:
https://docs.google.com/document/d/18nJsj-F2C_QXO7ClwzPcAUENQ-B440B43W7DdDnlDt4/edit?tab=t.0#heading=h.pgebnyi7pocj

But TLDR; 
guard_else_true, guard_else_false are better than guard_size_oblivious due to :
1) easier to reason about what assumptions we are making while reading the code. 
2) avoid size_oblivious complexity that is not needed. 
3) avoid unsoundness that could make guard_size_oblivious(a==1) be True when its not True for some a during runtime. 
4) less data dependent errors for some cases: ex, when  guard_size_oblivious(a==1)  we know a is a tensor size, but its traced with a=u1-u2 for example 
guard_size_oblivious(a==1)  will throw data dependent error but guard_else_false will just return false. 

cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Introduce guard_or_true, guard_or_false and avoid guard_size_oblivious in decompositions.py"


some context in this document:
https://docs.google.com/document/d/18nJsj-F2C_QXO7ClwzPcAUENQ-B440B43W7DdDnlDt4/edit?tab=t.0#heading=h.pgebnyi7pocj

But TLDR;
guard_or_true, guard_or_false are better than guard_size_oblivious due to :
easier to reason about what assumptions we are making while reading the code.
avoid size_oblivious complexity that is not needed.
avoid unsoundness that could make guard_size_oblivious(a==1) be True when its not True for some a during runtime.
less data dependent errors for some cases: ex, when guard_size_oblivious(a==1) we know a is a tensor size, but its traced with a=u1-u2 for example guard_size_oblivious(a==1) will throw data dependent error but guard_else_false will just return false.

**How is it different from statically_known_true??**

**statically_known_true:** would be used when you never want to add a guard, but just want to do a best effort check to see if you can infer that something is true ONLY based on existing constraints.

**guard_or_false/guard_or_false:** Those would be used in situations you prefer to guard and know the result of the expression over not guarding, but in case you hit a data dependent error you are ok with just returning true or false.
The reason you are ok with returning true/false instead could be:
1. It's an optimization I do not want to fail for not performing optimization.
2. I am willing to deviate from the normal semantics when I have unbacked for the benefit of not failing (See the doc above) for more details.



cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented on torch/_inductor/decomposition.py:
```diff
@@ -287,16 +287,14 @@ def addmm(
-            guard_size_oblivious(mat1.size(0) == 1)
+            guard_or_false(mat1.size(0) == 1)
             and definitely_true(mat2.size(0) <= 16)
             and definitely_true(mat2.size(1) <= 16)
```
maybe use guard_or_false here also?

aorenste commented:
#### Bikeshedding:
What if instead of two APIs (`guard_or_true`/`guard_or_false`) we have one that you specify the "unguarded" value? Something like `guard_or_else(expr, unguarded_value)`?

So current uses of `guard_or_false(x)` would be `guard_or_else(x, False)`

aorenste commented on torch/_inductor/decomposition.py:
```diff
@@ -261,15 +265,11 @@ def bmm(
-        if guard_size_oblivious(self.shape[1] == 1) or guard_size_oblivious(
-            batch2.shape[2] == 1
-        ):
+        if guard_or_false(self.shape[1] == 1) or guard_or_false(batch2.shape[2] == 1):
```
I've been thinking about this line (and examples like it) and what behavior we ACTUALLY want. Do we REALLY want something like:
```
if statically_known_true(self.shape[1] == 1 or batch2.shape[2] == 1):
    guard_or_false(self.shape[1] == 1) or guard_or_false(batch2.shape[2] == 1)
```
In other words - if we can determine this using the currently known information then do so and guard on whatever we needed otherwise don't add ANY guards.

laithsakka replied:
if we already know the value statically , then there is no value to guard right? guard wont guard in fact.

laithsakka committed:
Update on "Introduce guard_or_true, guard_or_false and avoid guard_size_oblivious in decompositions.py"


some context in this document:
https://docs.google.com/document/d/18nJsj-F2C_QXO7ClwzPcAUENQ-B440B43W7DdDnlDt4/edit?tab=t.0#heading=h.pgebnyi7pocj

But TLDR;
`guard_or_true`, `guard_or_false` are better than `guard_size_oblivious` due to :
- Easier to reason about what assumptions we are making while reading the code.
- Avoid size_oblivious complexity that is not needed.
- Avoid unsoundness that could make `guard_size_oblivious(a==1)` be true when its not true for some vaue `a` during runtime.
- Less data dependent errors for some cases: ex, when doing `guard_size_oblivious(a==1)` and we know `a` is a tensor size, if it's traced with `a=u1-u2` `guard_size_oblivious(a==1)` will throw a data dependent error but `guard_else_false` will just return `False`.

### How is it different from statically_known_true??
**`if(cond)`:** (normal guarding) will try to evaluate statically and guard on the condition, willing to restrict input space to evaluate cond. if it fails to evaluate due to data dependent error will throw an exception (that could be converted to graph break in some situations). 

**`statically_known_true(cond)`:** would be used when you never want to add a guard (restrict your input space), but just want to do a best effort check to see if you can infer that something is true/false ONLY based on existing constraints.

**`guard_or_true(cond)`/`guard_or_false(cond)`:** Those would be used in situations you prefer to guard and know the result of the expression over not guarding, but in case you hit a data dependent error you are ok with just returning true or false.
Some reasons you might be ok with returning true/false instead could be:
1. It's an optimization I do not want to fail for not performing optimization.
2. I am willing to deviate from the normal semantics when I have unbacked for the benefit of not failing (See the doc above for more details).

**`definitely_true(cond)`**: same as `guard_or_false(cond)` except does not try to do static eval for unbacked (planning to deprecate it and replace uses with `guard_or_false` or make it alias to `guard_or_false`)
 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "[DRAFT] Introduce guard_or_true, guard_or_false"


some context in this document:
https://docs.google.com/document/d/18nJsj-F2C_QXO7ClwzPcAUENQ-B440B43W7DdDnlDt4/edit?tab=t.0#heading=h.pgebnyi7pocj

But TLDR;
`guard_or_true`, `guard_or_false` are better than `guard_size_oblivious` due to :
- Easier to reason about what assumptions we are making while reading the code.
- Avoid size_oblivious complexity that is not needed.
- Avoid unsoundness that could make `guard_size_oblivious(a==1)` be true when its not true for some vaue `a` during runtime.
- Less data dependent errors for some cases: ex, when doing `guard_size_oblivious(a==1)` and we know `a` is a tensor size, if it's traced with `a=u1-u2` `guard_size_oblivious(a==1)` will throw a data dependent error but `guard_else_false` will just return `False`.

### How is it different from statically_known_true??
**`if(cond)`:** (normal guarding) will try to evaluate statically and guard on the condition, willing to restrict input space to evaluate cond. if it fails to evaluate due to data dependent error will throw an exception (that could be converted to graph break in some situations). 

**`statically_known_true(cond)`:** would be used when you never want to add a guard (restrict your input space), but just want to do a best effort check to see if you can infer that something is true/false ONLY based on existing constraints.

**`guard_or_true(cond)`/`guard_or_false(cond)`:** Those would be used in situations you prefer to guard and know the result of the expression over not guarding, but in case you hit a data dependent error you are ok with just returning true or false.
Some reasons you might be ok with returning true/false instead could be:
1. It's an optimization I do not want to fail for not performing optimization.
2. I am willing to deviate from the normal semantics when I have unbacked for the benefit of not failing (See the doc above for more details).

**`definitely_true(cond)`**: same as `guard_or_false(cond)` except does not try to do static eval for unbacked (planning to deprecate it and replace uses with `guard_or_false` or make it alias to `guard_or_false`)
 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "[DRAFT] Introduce guard_or_true, guard_or_false"


some context in this document:
https://docs.google.com/document/d/18nJsj-F2C_QXO7ClwzPcAUENQ-B440B43W7DdDnlDt4/edit?tab=t.0#heading=h.pgebnyi7pocj

But TLDR;
`guard_or_true`, `guard_or_false` are better than `guard_size_oblivious` due to :
- Easier to reason about what assumptions we are making while reading the code.
- Avoid size_oblivious complexity that is not needed.
- Avoid unsoundness that could make `guard_size_oblivious(a==1)` be true when its not true for some vaue `a` during runtime.
- Less data dependent errors for some cases: ex, when doing `guard_size_oblivious(a==1)` and we know `a` is a tensor size, if it's traced with `a=u1-u2` `guard_size_oblivious(a==1)` will throw a data dependent error but `guard_else_false` will just return `False`.

### How is it different from statically_known_true??
**`if(cond)`:** (normal guarding) will try to evaluate statically and guard on the condition, willing to restrict input space to evaluate cond. if it fails to evaluate due to data dependent error will throw an exception (that could be converted to graph break in some situations). 

**`statically_known_true(cond)`:** would be used when you never want to add a guard (restrict your input space), but just want to do a best effort check to see if you can infer that something is true/false ONLY based on existing constraints.

**`guard_or_true(cond)`/`guard_or_false(cond)`:** Those would be used in situations you prefer to guard and know the result of the expression over not guarding, but in case you hit a data dependent error you are ok with just returning true or false.
Some reasons you might be ok with returning true/false instead could be:
1. It's an optimization I do not want to fail for not performing optimization.
2. I am willing to deviate from the normal semantics when I have unbacked for the benefit of not failing (See the doc above for more details).

**`definitely_true(cond)`**: same as `guard_or_false(cond)` except does not try to do static eval for unbacked (planning to deprecate it and replace uses with `guard_or_false` or make it alias to `guard_or_false`)
 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Introduce guard_or_true, guard_or_false"


some context in this document:
https://docs.google.com/document/d/18nJsj-F2C_QXO7ClwzPcAUENQ-B440B43W7DdDnlDt4/edit?tab=t.0#heading=h.pgebnyi7pocj

But TLDR;
`guard_or_true`, `guard_or_false` are better than `guard_size_oblivious` due to :
- Easier to reason about what assumptions we are making while reading the code.
- Avoid size_oblivious complexity that is not needed.
- Avoid unsoundness that could make `guard_size_oblivious(a==1)` be true when its not true for some vaue `a` during runtime.
- Less data dependent errors for some cases: ex, when doing `guard_size_oblivious(a==1)` and we know `a` is a tensor size, if it's traced with `a=u1-u2` `guard_size_oblivious(a==1)` will throw a data dependent error but `guard_else_false` will just return `False`.

### How is it different from statically_known_true??
**`if(cond)`:** (normal guarding) will try to evaluate statically and guard on the condition, willing to restrict input space to evaluate cond. if it fails to evaluate due to data dependent error will throw an exception (that could be converted to graph break in some situations). 

**`statically_known_true(cond)`:** would be used when you never want to add a guard (restrict your input space), but just want to do a best effort check to see if you can infer that something is true/false ONLY based on existing constraints.

**`guard_or_true(cond)`/`guard_or_false(cond)`:** Those would be used in situations you prefer to guard and know the result of the expression over not guarding, but in case you hit a data dependent error you are ok with just returning true or false.
Some reasons you might be ok with returning true/false instead could be:
1. It's an optimization I do not want to fail for not performing optimization.
2. I am willing to deviate from the normal semantics when I have unbacked for the benefit of not failing (See the doc above for more details).

**`definitely_true(cond)`**: same as `guard_or_false(cond)` except does not try to do static eval for unbacked (planning to deprecate it and replace uses with `guard_or_false` or make it alias to `guard_or_false`)
 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1180,6 +1182,38 @@ def compute_unbacked_bindings(
+# Some reasons you might be ok with returning true/false instead could be:
+#  (1) It's an optimization/additional check I do not want to fail for not performing it.
+#  (2) I am willing to deviate from the normal semantics when I have unbacked for the
+#      benefit of not failing (See the doc above for more details).
```
remove See the doc above for more details before landing this

aorenste commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1180,6 +1182,38 @@ def compute_unbacked_bindings(
+#      benefit of not failing (See the doc above for more details).
+def guard_or_false(a: BoolLikeType) -> bool:
+    """
+    Try to gaurd a, if data dependent error encountered just return false.
```
nit: "guard" (and in `guard_or_true` too)

laithsakka replied:
addressed in https://github.com/pytorch/pytorch/pull/150085

aorenste commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1180,6 +1182,38 @@ def compute_unbacked_bindings(
+    """
+    Try to gaurd a, if data dependent error encountered just return false.
+    """
+    if isinstance(a, SymBool):
+        try:
+            guard_bool(a)
+        except GuardOnDataDependentSymNode:
+            return False
+    return bool(a)
```
guard_bool() is already doing all this work - maybe we should just use it:
```suggestion
      try:
          return bool(guard_bool(a))
      except GuardOnDataDependentSymNode:
          return False
```

laithsakka replied:
addressed in https://github.com/pytorch/pytorch/pull/150085

laithsakka committed:
Update on "Introduce guard_or_true, guard_or_false"


some context in this document:
https://docs.google.com/document/d/18nJsj-F2C_QXO7ClwzPcAUENQ-B440B43W7DdDnlDt4/edit?tab=t.0#heading=h.pgebnyi7pocj

But TLDR;
`guard_or_true`, `guard_or_false` are better than `guard_size_oblivious` due to :
- Easier to reason about what assumptions we are making while reading the code.
- Avoid size_oblivious complexity that is not needed.
- Avoid unsoundness that could make `guard_size_oblivious(a==1)` be true when its not true for some vaue `a` during runtime.
- Less data dependent errors for some cases: ex, when doing `guard_size_oblivious(a==1)` and we know `a` is a tensor size, if it's traced with `a=u1-u2` `guard_size_oblivious(a==1)` will throw a data dependent error but `guard_else_false` will just return `False`.

### How is it different from statically_known_true??
**`if(cond)`:** (normal guarding) will try to evaluate statically and guard on the condition, willing to restrict input space to evaluate cond. if it fails to evaluate due to data dependent error will throw an exception (that could be converted to graph break in some situations). 

**`statically_known_true(cond)`:** would be used when you never want to add a guard (restrict your input space), but just want to do a best effort check to see if you can infer that something is true/false ONLY based on existing constraints.

**`guard_or_true(cond)`/`guard_or_false(cond)`:** Those would be used in situations you prefer to guard and know the result of the expression over not guarding, but in case you hit a data dependent error you are ok with just returning true or false.
Some reasons you might be ok with returning true/false instead could be:
1. It's an optimization I do not want to fail for not performing optimization.
2. I am willing to deviate from the normal semantics when I have unbacked for the benefit of not failing (See the doc above for more details).

**`definitely_true(cond)`**: same as `guard_or_false(cond)` except does not try to do static eval for unbacked (planning to deprecate it and replace uses with `guard_or_false` or make it alias to `guard_or_false`)
 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
@pytorchbot commit

laithsakka commented:
@pytorchbot merge

laithsakka commented:
@pytorchmergebot revert

laithsakka commented:
@pytorchmergebot revert -m "need to fix stuff" -c weird

laithsakka committed:
Update on "Introduce guard_or_true, guard_or_false"


some context in this document:
https://docs.google.com/document/d/18nJsj-F2C_QXO7ClwzPcAUENQ-B440B43W7DdDnlDt4/edit?tab=t.0#heading=h.pgebnyi7pocj

But TLDR;
`guard_or_true`, `guard_or_false` are better than `guard_size_oblivious` due to :
- Easier to reason about what assumptions we are making while reading the code.
- Avoid size_oblivious complexity that is not needed.
- Avoid unsoundness that could make `guard_size_oblivious(a==1)` be true when its not true for some vaue `a` during runtime.
- Less data dependent errors for some cases: ex, when doing `guard_size_oblivious(a==1)` and we know `a` is a tensor size, if it's traced with `a=u1-u2` `guard_size_oblivious(a==1)` will throw a data dependent error but `guard_else_false` will just return `False`.

### How is it different from statically_known_true??
**`if(cond)`:** (normal guarding) will try to evaluate statically and guard on the condition, willing to restrict input space to evaluate cond. if it fails to evaluate due to data dependent error will throw an exception (that could be converted to graph break in some situations). 

**`statically_known_true(cond)`:** would be used when you never want to add a guard (restrict your input space), but just want to do a best effort check to see if you can infer that something is true/false ONLY based on existing constraints.

**`guard_or_true(cond)`/`guard_or_false(cond)`:** Those would be used in situations you prefer to guard and know the result of the expression over not guarding, but in case you hit a data dependent error you are ok with just returning true or false.
Some reasons you might be ok with returning true/false instead could be:
1. It's an optimization I do not want to fail for not performing optimization.
2. I am willing to deviate from the normal semantics when I have unbacked for the benefit of not failing (See the doc above for more details).

**`definitely_true(cond)`**: same as `guard_or_false(cond)` except does not try to do static eval for unbacked (planning to deprecate it and replace uses with `guard_or_false` or make it alias to `guard_or_false`)
 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge

laithsakka commented:
@pytorchbot commit -i

laithsakka commented:
@pytorchbot merge -i

laithsakka commented:
@pytorchbot merge

laithsakka commented:
@pytorchbot commit

laithsakka commented:
pytorchbot merge

laithsakka commented:
alreaded landed
